### ISSUE TITLE
Thread Cleanup Infinite Loop

### ISSUE THREAD

**Issue by neldridge** [jenkinsci/openstack-cloud-plugin#171]
First, thank you for your work on this plugin- we utilize it heavily.

Plugin version: 2.23
Jenkins version: 2.73.1
OpenStack version: kilo 

We have an ongoing issue where the thread cleanup will get into an infinite loop, executors won't be cleaned up, new executors won't start due to this, and resolution is to restart Jenkins.

Logs from our Jenkins container here: https://gist.github.com/neldridge/9a303f29b902e3a3d1bde04bc55142b1

I think the core of the issue is with our OpenStack implementation; we have other services that will sometimes hang forever waiting on the response from OS (Terraform sometimes gets stuck refreshing state of a resource); we are actively working on resolving this.

Even with this issue, I believe the plugin should be able to handle cleanup better; if the loop is running past X, restart the thread.

Please let me know if there's anything further you need from me, or if I can help in some way.

**Comment by olivergondza** [comment]
Strange, can you capture `jstack -l $PID` at the time thread named "OpenStack slave cleanup" is doing something? If you are getting different stacktraces on repeated attempts, please bring them all. Without that, we have no idea why it takes so long.

**Comment by imriz** [comment]
We suffer from this too.

Plugin version: 2.24
Jenkins version: 2.24
OpenStack version: Liberty

It should be noted that we believe this started after upgrading Jenkins (from 2.46) and the plugin (from 2.22).

We will try to downgrade and report.

As for the cleanup thread - It seems that killing the thread (via the /monitoring endpoint) gets everything back in motion, thus avoiding restart.

This is the thread stacktrace when the problem happens.
```
OpenStack slave cleanup thread
sun.misc.Unsafe.park(Native Method)
java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
org.apache.http.pool.PoolEntryFuture.await(PoolEntryFuture.java:138)
org.apache.http.pool.AbstractConnPool.getPoolEntryBlocking(AbstractConnPool.java:306)
org.apache.http.pool.AbstractConnPool.access$000(AbstractConnPool.java:64)
org.apache.http.pool.AbstractConnPool$2.getPoolEntry(AbstractConnPool.java:192)
org.apache.http.pool.AbstractConnPool$2.getPoolEntry(AbstractConnPool.java:185)
org.apache.http.pool.PoolEntryFuture.get(PoolEntryFuture.java:107)
org.apache.http.impl.conn.PoolingHttpClientConnectionManager.leaseConnection(PoolingHttpClientConnectionManager.java:276)
org.apache.http.impl.conn.PoolingHttpClientConnectionManager$1.get(PoolingHttpClientConnectionManager.java:263)
org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:190)
org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
org.openstack4j.connectors.httpclient.HttpCommand.execute(HttpCommand.java:112)
org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.invokeRequest(HttpExecutorServiceImpl.java:65)
org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.invoke(HttpExecutorServiceImpl.java:56)
org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.execute(HttpExecutorServiceImpl.java:32)
org.openstack4j.core.transport.internal.HttpExecutor.execute(HttpExecutor.java:51)
org.openstack4j.openstack.internal.BaseOpenStackService$Invocation.execute(BaseOpenStackService.java:208)
org.openstack4j.openstack.internal.BaseOpenStackService$Invocation.execute(BaseOpenStackService.java:202)
org.openstack4j.openstack.compute.internal.ServerServiceImpl.list(ServerServiceImpl.java:97)
org.openstack4j.openstack.compute.internal.ServerServiceImpl.list(ServerServiceImpl.java:82)
jenkins.plugins.openstack.compute.internal.Openstack.getRunningNodes(Openstack.java:218)
jenkins.plugins.openstack.compute.JCloudsCleanupThread.destroyServersOutOfScope(JCloudsCleanupThread.java:161)
jenkins.plugins.openstack.compute.JCloudsCleanupThread.execute(JCloudsCleanupThread.java:68)
hudson.model.AsyncPeriodicWork$1.run(AsyncPeriodicWork.java:101)
java.lang.Thread.run(Thread.java:748)
```

**Comment by svijee** [comment]
> As for the cleanup thread - It seems that killing the thread (via the /monitoring endpoint) gets everything back in motion, thus avoiding restart.

Can you tell me how I can do this exactly?

**Comment by imriz** [comment]
Sure. Assuming Jenkins 2, Go to and expand the threads list: 
https://<Your Jenkins URL>/monitoring#threads

Look for the cleanup thread, and click the red button on the right (icon of a red flag).

**Comment by imriz** [comment]
Could it be that 73d4f8c0f613983012f01d532a1bfa7a470a126d is the root cause?
From first glance, it seems it is caching the connection? Guesstimate ahead, but maybe openstack4j doesn't set timeout by default, and by caching the connection we increase the chance of some weird connection issues? Maybe related to ContainX/openstack4j/issues/846?

**Comment by olivergondza** [comment]
I knew it sounds familiar: https://github.com/ContainX/openstack4j/issues/579#issuecomment-319641117. FTR, the okhttp i gave a try did not worked much better, I found it started hundreds pooled connection threads that did not went away easily.

**Comment by olivergondza** [comment]
I will switch to ok-http for next release to fix the underlying cause here - the connection can get stuck blocking dependent actions.

**Comment by neldridge** [comment]
I finally got a stack trace, it hadn't happened til over the weekend. Not sure if it's useful since it sounds like you've found the issue but I'll attach anyway.

Full dump here:  https://gist.github.com/neldridge/f224e467a47d7c9045269ec31aa5066e

```

"OpenStack slave cleanup thread" daemon prio=5 RUNNABLE
	java.net.SocketInputStream.socketRead0(Native Method)
	java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	java.net.SocketInputStream.read(SocketInputStream.java:171)
	java.net.SocketInputStream.read(SocketInputStream.java:141)
	sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
	sun.security.ssl.InputRecord.read(InputRecord.java:503)
	sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:983)
	sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1385)
	sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1413)
	sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1397)
	org.apache.http.conn.ssl.SSLConnectionSocketFactory.createLayeredSocket(SSLConnectionSocketFactory.java:394)
	org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:353)
	org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:141)
	org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
	org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
	org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
	org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
	org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
	org.openstack4j.connectors.httpclient.HttpCommand.execute(HttpCommand.java:112)
	org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.invokeRequest(HttpExecutorServiceImpl.java:65)
	org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.invoke(HttpExecutorServiceImpl.java:56)
	org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.execute(HttpExecutorServiceImpl.java:32)
	org.openstack4j.core.transport.internal.HttpExecutor.execute(HttpExecutor.java:51)
	org.openstack4j.openstack.internal.OSAuthenticator.authenticateV2(OSAuthenticator.java:122)
	org.openstack4j.openstack.internal.OSAuthenticator.invoke(OSAuthenticator.java:52)
	org.openstack4j.openstack.client.OSClientBuilder$ClientV2.authenticate(OSClientBuilder.java:117)
	org.openstack4j.openstack.client.OSClientBuilder$ClientV2.authenticate(OSClientBuilder.java:79)
	jenkins.plugins.openstack.compute.internal.Openstack.<init>(Openstack.java:121)
	jenkins.plugins.openstack.compute.JCloudsCloud.getOpenstack(JCloudsCloud.java:456)
	jenkins.plugins.openstack.compute.JCloudsCleanupThread.destroyServersOutOfScope(JCloudsCleanupThread.java:161)
	jenkins.plugins.openstack.compute.JCloudsCleanupThread.execute(JCloudsCleanupThread.java:68)
	hudson.model.AsyncPeriodicWork$1.run(AsyncPeriodicWork.java:101)
	java.lang.Thread.run(Thread.java:748)
```

**Comment by olivergondza** [comment]
@neldridge, the threaddump you have attached looks quite healthy to me: 2 threads, one opening ssl session, the other reading the response. What are the undesired effects aside of what is observed in threaddump.

But as you said this is likely to go away with the update.

**Comment by imriz** [comment]
@olivergondza I usually don't like to guess, but I'll bite anyway - Could it be that openstack4j doesn't set a timeout, and so the symptoms described in this issue could also happen from connection issue which will result a request being stuck forever?

**Comment by olivergondza** [comment]
@imriz, that is one of the possibilities. But it really depends on what do we actually observe, it is either an infinite loop and in that case the thread is certainly not stuck reading the socket or it is stuck but then it is not an infinite loop as reported. There is no way to tell where we are from the single threaddump or without poking around live impacted instance.

**Comment by imriz** [comment]
@olivergondza yep, agreed. Let's hope the move to okhttp resolved this.
Awesome plugin, btw :)

**Comment by svijee** [comment]
Thanks for the update @olivergondza!

However, we still have issues. After the update and waiting for a couple of hours, some machine spin up, some don't. For example, one pipeline failed with:

`[ubuntu1604-64-cloud] ERROR: Issue with creating launcher for agent ubuntu1604-64-cloud-9788. The agent has not been fully initialized yet`

Relevant log:
`Sep 27, 2017 5:41:24 AM okhttp3.ConnectionPool pruneAndGetAllocationCount
WARNING: A connection to http://$OPENSTACKURL:8774/ was leaked. Did you forget to close a response body?`

The last message appeared a couple of times in the jenkins log. Not sure if the others in this issue have similar problems.

**Comment by olivergondza** [comment]
@svijee, there seem to be 2 problems you are running into. The okhttp connection leak report will be fixed in https://github.com/jenkinsci/openstack-cloud-plugin/commit/ef1a9df7397c3e53de6db15731e6349dfc5b1d26.

The problem with the channel closed does not seem to be related to this ticket at all. Consult the log from the agent JVM to see what is going on.

**Comment by svijee** [comment]
@olivergondza Thanks! Looking into it!

**Comment by neldridge** [comment]
@olivergondza i can't speak to the health status other than the build queue was backed up due to executor starvation, and the OS executors were "disconnected" but not being cleaned up... They were also still live in OS.

I restarted Jenkins, and it cleaned it self up a few minutes later.

I've updated to 2.25.. 

If it happens again, I'll try to compile more information.. Thanks again for your all your hard work.
